core/src/main/java/feign/Expander.java

```diff
@@ -0,0 +1,33 @@
+/**
+ * Copyright 2012-2018 The Feign Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License
+ * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+ * or implied. See the License for the specific language governing permissions and limitations under
+ * the License.
+ */
+package feign;
+
+/**
+ * Defines how to expand the value of a query parameter.
+ */
+public interface Expander {
+
+  /**
+   * Expands the value into a string. Does not accept or return null.
+   */
+  String expand(Object value);
+
+  final class ToStringExpander implements Expander {
```

Comment:
I think a more appropriate name here is `StringExpander`, or `DefaultExpander`.  This reveals its intention more clearly.

Reply:
@kdavisk6 This was pulled from `Param.ToStringExpander` so the naming just got copied over. It is the default (so `DefaultExpander` isn't a bad name) and it expands by calling `value.toString()` (therefore, IMO, `ToStringExpander` is a more descriptive name). Or do you mean that the `Expander` interface should be renamed to `StringExpander`?

Reply:
First, it is best to avoid the use of verbs in class names.  It's preferable to use noun or noun phrase names, as Classes represent objects not actions performed.  That is the responsibility of Methods.

Second, following the principle of using "intention revealing names", the name `ToStringExpander` doesn't clearly state it's intention at first glance.  The use of the preposition "To" may indicate that there is a "From" version elsewhere, providing false information that this object may be part of a conversion chain.  Only upon inspecting the implementation that we find it's not.

Pulling on that thread further, the intention of the interface `Expander` seems at odds with it's intended use.  If I understand your intent, this interface is meant to be a bridge, allowing users to have finer control over how a `QueryMap` annotated value is represented on a `Request`.  Another option for the interface name could be `QueryMapConverter`, which identifies that the purpose of the object is to take a `QueryMap` value and `convert` it into something else.  

```java
public interface QueryMapConverter {
     public String convert(Object value);
}

/**
 * QueryMap Converter which maps all values to their String representation
 */
public class StringMappingQueryMapConverter implements QueryMapConverter {
      @Override
      public String convert(Object value) {
         return value.toString();
      }
}
```

This results in a design that that is clear from first glance and makes the design of the system explicit.  Adding new converters becomes a straight forward task.  Thoughts?

Reply:
@kdavisk6 I don't think this has anything to do with a `QueryMap`. It is just an `Expander` (see `Param.Expander` and `Param.ToStringExpander` -- these names already exist in the master branch) which (in this context) is being used to serialize the values from the query map. It still follows everything from the `Expander` interface, never knowing whether it's values came from a `Map` or from a `Param` annotated parameter. As far as "finer control", I think this is just intended to allow the user to specify that a `QueryMap` may use an expander other than the default `Param.ToStringExpander` (which is essentially what's happening now). Personally, I think a better solution is to use POJOs and annotate the POJO parameters so that you can have control over each value's expander individually, but that's not what this ticket addresses.

Also, I disagree about `ToStringExpander` being an inherently bad name (for reasons mentioned earlier). It is not a verb, it is descriptive, and when I saw the name in my first pass through the Feign code, I immediately guessed correctly what it did. However, if `ToStringExpander` seems easy to misread to you, then perhaps `DefaultExpander` would be better (less chance of confusion at least, even if it is less descriptive).

I don't like the current `Expander` (param serializer) vs `Encoder` (body serializer) naming. How are you supposed to know what the difference is without reading the documentation? So I do think that at least `QueryMapConverter` tells you something about where it is used, even if it is misleading (it doesn't convert a query map, it just converts individual values -- so maybe `QueryMapValueConverter` would be better).

Reply:
I will go back to my favorite quote from *Clean Code*:

> You know you are working on clean code when each routine you read turns out to be pretty much what you expected.  You can call it beautiful code when the code also makes it look like the language was made for the problem -- Ward Cunningham

Looking at our discussion, the code we are reviewing does not fit this particular definition.  This is not an admonishment of the contributor, nor a critique of the existing approach of which this change is based.  It's an observation that, if it did fit this definition, we would be more concerned with the intent of the work and not the name of the objects.  We should always be looking to follow the Boy Scout Rule:

> Leave the campground cleaner than you found it.

Those not from the United States may not get that reference, but the spirit of the quote is the same.  It means that we should strive to refactor and remove ambiguity from our code as preventative maintenance, with an eye towards knowing that others should be able to understand our intent long after we are gone.  *Leave things better than we found it.*  Just because it has been done this way doesn't mean we can't change it.

Case in point, our confusion over the initial interface `Expander`.   Those who came before us had a mental model of how all of these objects should interact.  Their choice of naming fit that model however, the clarity they shared when building these components is missing from the rest of the design.

To bring this back to my initial comment, I still feel that the name of the components both `Expander` and `ToStringExpander` do not properly express their intended use, regardless of the components history or genesis.    Doing a search for where `Expander` is used in the rest of the library and how it is being used indicates that it's purpose is to `Convert`.  

https://github.com/OpenFeign/feign/blob/9fd1b1dd31dc29834b293edfbd16a967fa746a89/core/src/test/java/feign/ContractWithRuntimeInjectionTest.java#L35-L52

In this example it's used to convert values into lower case.  And in `ReflectiveFeign`

https://github.com/OpenFeign/feign/blob/5e2fdeed0dc72cf7022993ad9877f7bb8e011e14/core/src/main/java/feign/ReflectiveFeign.java#L209-L221

It's used to convert the request parameters before submitting the request.  With that I suggest that the interface `Expander` should be changed to `Converter`, with the interface method also changed to `convert` and that the default implementation of the `Converter` be what is already in place.

```java
public interface Converter {
   public String convert(Object value);
}
```

Reply:
@velo any thoughts?

Reply:
I don't know that "converer" is the right word, but even if it's not perfect, it's still a decent name. What it does is serialize (or "convert" I guess) query parameter objects into strings (which may be pre-encoded, or may still need encoding).

Reply:
`Converter`/`Expander`, both look same to me....

Lemme throw more options, what about `ParameterSerializer`?

Reply:
@rage-shadowman how does it feel?

Reply:
Lol, so is there a clear directive here for branch owner? Such as: rename `Expander` to `ParameterSerializer` and `ToStringExpander` to `ParameterSerializer.Default`?